src/ripple/protocol/impl/Feature.cpp

```diff
@@ -153,45 +168,92 @@ uint256 bitsetIndexToFeature(size_t i)
+uint256 const featureChecks = getMandatoryFeature("Checks");
+uint256 const fix1571 = getMandatoryFeature("fix1571");
+uint256 const fix1543 = getMandatoryFeature("fix1543");
+uint256 const fix1623 = getMandatoryFeature("fix1623");
+uint256 const featureDepositPreauth = getMandatoryFeature("DepositPreauth");
+uint256 const fix1515 = getMandatoryFeature("fix1515");
+uint256 const fix1578 = getMandatoryFeature("fix1578");
+uint256 const featureMultiSignReserve = getMandatoryFeature("MultiSignReserve");
+uint256 const fixTakerDryOfferRemoval = getMandatoryFeature("fixTakerDryOfferRemoval");
+uint256 const fixMasterKeyAsRegularKey = getMandatoryFeature("fixMasterKeyAsRegularKey");
+uint256 const fixCheckThreading = getMandatoryFeature("fixCheckThreading");
+uint256 const fixPayChanRecipientOwnerDir = getMandatoryFeature("fixPayChanRecipientOwnerDir");
+uint256 const featureDeletableAccounts = getMandatoryFeature("DeletableAccounts");
+uint256 const fixQualityUpperBound = getMandatoryFeature("fixQualityUpperBound");
+uint256 const featureRequireFullyCanonicalSig = getMandatoryFeature("RequireFullyCanonicalSig");
+uint256 const fix1781 = getMandatoryFeature("fix1781");
+uint256 const featureRetire2017Amendments = getMandatoryFeature("Retire2017Amendments");
 
 // The following amendments have been active for at least two years.
 // Their pre-amendment code has been removed.
-uint256 const retiredPayChan = *getRegisteredFeature("PayChan");
-uint256 const retiredCryptoConditions = *getRegisteredFeature("CryptoConditions");
-uint256 const retiredTickSize = *getRegisteredFeature("TickSize");
-uint256 const retiredFix1368 = *getRegisteredFeature("fix1368");
-uint256 const retiredEscrow = *getRegisteredFeature("Escrow");
-uint256 const retiredFix1373 = *getRegisteredFeature("fix1373");
-uint256 const retiredEnforceInvariants = *getRegisteredFeature("EnforceInvariants");
-uint256 const retiredSortedDirectories = *getRegisteredFeature("SortedDirectories");
-uint256 const retiredFix1201 = *getRegisteredFeature("fix1201");
-uint256 const retiredFix1512 = *getRegisteredFeature("fix1512");
-uint256 const retiredFix1523 = *getRegisteredFeature("fix1523");
-uint256 const retiredFix1528 = *getRegisteredFeature("fix1528");
+//
+// The static retired amendments could hypothetically be moved so they are only
+// inside the detail::retiringAmendments() implementation.  However doing so
+// would postpone the discovery of any construction problems until the first
+// call to retiringAmendments().  By leaving these definitions at file
+// scope any run-time build problems will be revealed before main() is called.
+static uint256 const retiredFeeEscalation = getMandatoryFeature("FeeEscalation");
+static uint256 const retiredMultiSign = getMandatoryFeature("MultiSign");
+static uint256 const retiredTrustSetAuth = getMandatoryFeature("TrustSetAuth");
+static uint256 const retiredFlow = getMandatoryFeature("Flow");
```

Comment:
A minor question: why the 4 retired* variable above? They did not exist in the last version. I thought it could be for documentation. But according to step 12 of the spec, they will be removed later.

Reply:
Those four `retired` variables are the first four features that I unconditionalized.  When I unconditionalized them I also removed the `feature` definitions because they were no longer referenced in the code base.  Then I started to research how to retire amendments and I discovered that I needed the values in order to retire them.  So I had to bring them back.

All of the `retired` features, including these four, are listed in the `retiringAmendments` list, below: https://github.com/ripple/rippled/pull/3368/files#diff-df6c689515140ea3ed30f8db7a9978b8R239-R242. That list is used to determine which amendments should be removed from the ledger and the `AmendmentTable`.

Reply:
Make sense. Thanks for the explanation!